br/pkg/lightning/backend/external/merge_v2.go

```diff
@@ -0,0 +1,186 @@
+package external
+
+import (
+	"bytes"
+	"context"
+	"math"
+	"time"
+
+	"github.com/jfcg/sorty/v2"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/br/pkg/membuf"
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"github.com/pingcap/tidb/pkg/util/size"
+	"go.uber.org/zap"
+)
+
+// MergeOverlappingFilesV2 reads from given files whose key range may overlap
+// and writes to new sorted, nonoverlapping files.
+// Using 1 readAllData and 1 writer.
+func MergeOverlappingFilesV2(
+	ctx context.Context,
+	dataFiles []string,
+	statFiles []string,
+	store storage.ExternalStorage,
+	startKey []byte,
+	endKey []byte,
+	partSize int64,
+	newFilePrefix string,
+	writerID string,
+	blockSize int,
+	writeBatchCount uint64,
+	propSizeDist uint64,
+	propKeysDist uint64,
+	onClose OnCloseFunc,
+	concurrency int,
+	checkHotspot bool,
+) (err error) {
+	task := log.BeginTask(logutil.Logger(ctx).With(
+		zap.Int("data-file-count", len(dataFiles)),
+		zap.Int("stat-file-count", len(statFiles)),
+		zap.Binary("start-key", startKey),
+		zap.Binary("end-key", endKey),
+		zap.String("new-file-prefix", newFilePrefix),
+		zap.Int("concurrency", concurrency),
+	), "merge overlapping files")
+	defer func() {
+		task.End(zap.ErrorLevel, err)
+	}()
+
+	rangesGroupSize := 4 * size.GB
+	failpoint.Inject("mockRangesGroupSize", func(val failpoint.Value) {
+		rangesGroupSize = uint64(val.(int))
+	})
+
+	splitter, err := NewRangeSplitter(
+		ctx,
+		dataFiles,
+		statFiles,
+		store,
+		int64(rangesGroupSize),
+		math.MaxInt64,
+		int64(4*size.GB),
+		math.MaxInt64,
+		checkHotspot,
+	)
+	if err != nil {
+		return err
```

Comment:
Should we print this error to log? otherwise, when the line 85 or 88 issues another error, the error will be overwriten?